Sprint 09

AlfaBotan opened the pull request:
Я так и не понял какую картинку нужно выбрать для профиля пользователя, добавил нейтральную, также  для наполнение профиля информацией, выбрал соответствующую смыслу лэйблов.
В остальном вроде бы всё в порядке)

AlfaBotan committed:
добавил UITabBarController и UINavigationController, через Segue реализован переход на экран с выбранным изображением, добавлен экран с профилем пользователя и реализован переход в него через UITabBarController.

AlfaBotan committed:
добавлен ScrollVuew на экран показа выбранного изображения, реализован функционал зума и автоматического растягивания изображения на весь экран с сохранением пропорций.
так же добавлена кнопка для шеринга выбранного изображения и реализована ее логика.

AlfaBotan committed:
Изменил реализацию постановки UI элементов на экран профиля пользователя
Все элементы добавлены через код.

Helvetii951 commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,106 @@
+    private var nameLable = UILabel()
+    private var descriptionLable = UILabel()
+    private var statusLable = UILabel()
+    private var exitbutton = UIButton()
```
Рекомендую всегда использовать [ленивые свойства ](https://www.avanderlee.com/swift/lazy-var-property/)для объявления подобных вьюх.  В нашем случае это какого-то буста не даст. Но, если какая-то вьюха/свойство не будет использоваться, можно сэкономить память.

Helvetii951 commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,106 @@
+   
+    @IBOutlet weak var logoutButton: UIButton!
+    
+    private var profileImage = UIImageView()
```
Такого аутлета в сториборде нет. Из-за этого приложение дальше упадет.

Helvetii951 commented on ImageFeed/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,74 @@
+    }
+    
+    @IBOutlet weak var backButton: UIButton!
+    @IBOutlet weak var sharingButton: UIButton!
```
Свойства можно сделать приватными. Рекомендую всегда писать функции и свойства с минимальной областью видимости, а затем менять ее, если нужно.

Helvetii951 commented on ImageFeed/SingleImageView/SingleImageViewController.swift:
```diff
@@ -0,0 +1,74 @@
+        scrollView.setContentOffset(CGPoint(x: x, y: y), animated: false)
+    }
+    
+    @IBAction func didTapShareButton() {
+        guard let image else { return }
+        let share = UIActivityViewController(
+            activityItems: [image],
+            applicationActivities: nil
+        )
+        present(share, animated: true, completion: nil)
+    }
+    @IBAction func didTapBackButton() {
+        dismiss(animated: true, completion: nil)
+    }
+}
```
Функции тоже можно сделать приватными.

AlfaBotan replied:
Ренат, добрый день)
Подскажите, пунк второй в критических замечаниях (Interface Builder выдает ошибки в верстке SingleImageViewController)
немного не понятно что конкретно не верно, если наставник ясно дал понять, что так и должно быть, да и в учебных материалах сказано следующее
<img width="2160" alt="Снимок экрана 2024-05-11 в 18 00 27" src="https://github.com/AlfaBotan/ImageFeed/assets/119239448/c046d017-af1b-40ce-9a32-d649f8a89f19">
<img width="397" alt="Снимок экрана 2024-05-11 в 18 00 46" src="https://github.com/AlfaBotan/ImageFeed/assets/119239448/e4a24b9b-22eb-4c46-a341-adb9b983ca7d">

так же просто для уточнения по третьему пункту, все отступы я брал из макета в фигме, но там два варианта макета и я работал с расширенной версией и в ней отступы больше чем в базовой версии. Я так понимаю, что нужно было брать отступы из базовой?

И что бы не возвращаться к вопросам)
Не в критических замечаниях, указано что изображение на SingleImageViewController, не выровнена по центру, не могли бы вы прикрепить скрин, что бы я точно понимал в какой момент вы это заметили?
Потому что при переходе на этот экран, картинка расположена ровно по центру о чём свидетельствует ползунок по горизонтали в  ScrollView в самом низу картинки, который указывает что можно скролить изображение влево и вправо, и там этот ползунок ровно по центру.

Буду благодарен если ответите.

Helvetii951 replied:
> работал с расширенной версией и в ней отступы больше чем в базовой версии

Не совсем. Верхний оступ там тоже меньше
<img width="560" alt="Снимок экрана 2024-05-11 в 20 18 57" src="https://github.com/AlfaBotan/ImageFeed/assets/4989481/ed672577-f418-487b-b103-8818baba9934">

Helvetii951 replied:
>Interface Builder выдает ошибки в верстке SingleImageViewController
Ну ок, тогда, закрываем. Видимо, я подзабыл учебник.

Helvetii951 replied:
>Interface Builder выдает ошибки в верстке SingleImageViewController

Ну ок, тогда, закрываем. Видимо, я немного подзабыл учебник.

Helvetii951 replied:
>Не в критических замечаниях, указано что изображение на SingleImageViewController, не выровнена по центру

В дизайне картинка занимает всю область экрана, в т.ч и залезает под SafeArea
<img width="192" alt="Снимок экрана 2024-05-11 в 20 25 18" src="https://github.com/AlfaBotan/ImageFeed/assets/4989481/3082bb76-197b-47d8-a35b-0a985830d28f">

В твоем контоллера есть отступ сверху. И он визуально как бы смещает картинку ниже центра. Но это только рекомендация к правкам.
![Simulator Screenshot - iPhone X - 2024-05-11 at 20 22 52](https://github.com/AlfaBotan/ImageFeed/assets/4989481/0be155fe-f8c7-4674-901e-a773ef8776a5)

AlfaBotan replied:
всё, до меня дошло, в макете отступы указаны до границ экрана, а не до safeArea)

AlfaBotan replied:
ну, я поправил эту рекомендацию, правда пришлось сделать не так ка в учебнике. В учебнике сказано, что скролл вью нужно привязать к safeArea с отступом 0, что соответственно не дает нам возможности показать картинку так как на макете в фигме.
Перепривязал отступы к SuperView)

AlfaBotan replied:
отправляю на перепроверку)

AlfaBotan committed:
Удалил Старый Аутлет кнопки и так же вызов его во вью дид лоад, что могло привести к падению приложения.
Изменил отступы элементов в профиле
Добавил недостающий приватный урвень доступа некоторым свойствам и методам
Сделал ленивыми свойства класса ProfileViewController для лейблов, картинки и кнопки.

AlfaBotan committed:
Изменил привязку ScrollView с SafeArea что бы полноценно реализовать макет дизайна в Фигме)
